api/pkg/artifact/artifact.go

```diff
@@ -69,14 +80,46 @@ func (gac *GcsArtifactClient) ParseURL(gsURL string) (*URL, error) {
 	}, nil
 }
 
+type Service interface {
+	GetType() string
+	GetURLScheme() string
+	ParseURL(gsURL string) (*URL, error)
+	ReadArtifact(ctx context.Context, url string) ([]byte, error)
+	WriteArtifact(ctx context.Context, url string, content []byte) error
+	DeleteArtifact(ctx context.Context, url string) error
+}
+
+type GcsArtifactClient struct {
+	URLScheme
+	api *storage.Client
+}
+
+func NewGcsArtifactClient() (Service, error) {
+	api, err := storage.NewClient(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed initializing gcs for the artifact client with error: %s", err.Error())
+	}
+	return &GcsArtifactClient{
+		URLScheme: gcsURLScheme,
+		api:       api,
+	}, nil
+}
+
+func (gac *GcsArtifactClient) GetType() string {
+	return gcsArtifactClientType
+}
+
 func (gac *GcsArtifactClient) ReadArtifact(ctx context.Context, url string) ([]byte, error) {
 	u, err := gac.ParseURL(url)
 	if err != nil {
 		return nil, err
 	}
 
-	reader, err := gac.API.Bucket(u.Bucket).Object(u.Object).NewReader(ctx)
+	reader, err := gac.api.Bucket(u.Bucket).Object(u.Object).NewReader(ctx)
```

Comment:
is it possible that `Object(u.Object)` return nil?

Reply:
Nope it's not possible for it to return nil or rather an empty string since an error would've already been returned by `ParseURL` if either `u.Bucket` or `u.Object` is empty: https://github.com/caraml-dev/mlp/pull/116/files#diff-944b4a0ae037725abba91bee6c62f431401c649ec9fbc3dc5e18271d88f253b1R69